Chgill-update-namespaces-design-doc (#11250)
NuGet/Home: implemented/2021/PackageSourceMapping.md

# Package Source Mapping

- [Jon Douglas](https://github.com/JonDouglas) & [Nikolche Kolev](https://github.com/nkolev92) & [Chris Gill](https://github.com/chgill-MSFT)
- Start Date: (2021-02-19)
- GitHub Issue ([6867](https://github.com/NuGet/Home/issues/6867))

> ## Status

Reviewer: Love this update :)

This feature is released in preview and will GA with Visual Studio 17.0.

Release announcement: https://devblogs.microsoft.com/nuget/introducing-package-source-mapping/

 compatible with the following tools:

* [Visual Studio 2022 preview 4](https://visualstudio.microsoft.com/vs/preview/) and later
* [nuget.exe 6.0.0-preview.4](https://www.nuget.org/downloads) and later
* [.NET SDK 6.0.100-rc.1](https://devblogs.microsoft.com/nuget/introducing-package-source-mapping/#package-source-mapping-rules) and later

Older tooling will ignore the Package Source Mapping configuration. To use this feature, ensure all your build environments use compatible tooling versions.

> Package Source Mappings will apply to all project types – including .NET Framework – as long as compatible tooling is used for build and restore.

Reviewer: nit: `including .NET Framework` seems unnecessary. Is there a guideline?

I usually try not to associate tooling with the framework explicitly, even if customers do it themselves.

Author: Removed the specific callout here as unnecessary🙂

## Summary

<!-- One-paragraph description of the proposal. -->

> NuGet allows access to all package IDs on a package source. However, many .NET developers have needs to filter & consume a subset of the package IDs available on public, private, and local sources. Although NuGet supports a concept of [trusted-signers](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-trusted-signers) to trust packages signed by specific authors, users would like to optionally specify which package IDs shoulder be allowed from which sources.

Reviewer: ```suggestion
NuGet allows access to all package IDs on a package source. However, many .NET developers have needs to filter & consume a subset of the package IDs available on public, private, and local sources. Although NuGet supports a concept of [trusted-signers](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-trusted-signers) to trust packages signed by specific authors, users would like to optionally specify which package IDs should be allowed from which sources.
```

This work will also allow future experiences in browsing, installing, and updating packages under an allowed list of package ID patterns to make developers secure by default when managing their dependencies in their software supply chain.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
When using a combination of public, private, and local sources defined in `NuGet.config` file(s), a user can add a new `<packageSourceMapping>` element to opt-in to the feature. Lastly by adding individual `<package pattern="">` elements in the `<packageSource>` node, the source will only allow the matching package IDs from the respective package source.

> **Prerequisites**

Reviewer: Sounds like most of these are recommendations rather than prereqs.

> 1. Create a repo-level `nuget.config` file by executing `dotnet new nugetconfig` at the repo root.

Reviewer: Have or create one is probably the better language here.

Author: Agree 👍

> 2. Define a [repo-specific global packages folder](https://docs.microsoft.com/nuget/reference/nuget-config-file#config-section) in your `nuget.config`.

Reviewer: This is probably merely a recommendation, rather than a requirement. 

I think calling out the global packages folder that frequently is unnecessary.

Author: Agree 👍

3. Define your package sources preceded with a `<clear />` to ensure no sources are inherited from a lower level config.

```xml
<!-- Define a global packages folder for your repository. -->
<!-- This is where installed packages will be stored locally. -->
<config>
  <add key="globalPackagesFolder" value="globalPackagesFolder" />
</config>

<!-- Define my package sources, nuget.org and contoso.com. -->
<!-- `clear` ensures no additional sources are inherited from another config file. -->
<packageSources>
  <clear />
  <!-- `key` can be any identifier for your source. -->
  <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
  <add key="contoso" value="https://contoso.com/packages/" />
</packageSources>
```

**Definition:**

Add a new `<packageSourceMapping>` element within the `NuGet.config` using the following syntax:

> ```xml

Reviewer: Seems like there's some duplication here with the global packages folder.

```xml
<!-- Define a global packages folder for your repository. -->
<!-- This is where installed packages will be stored locally. -->
<config>
  <add key="globalPackagesFolder" value="globalPackagesFolder" />
</config>

<!-- Define my package sources, nuget.org and contoso.com. -->
<!-- `clear` ensures no additional sources are inherited from another config file. -->
<packageSources>
  <clear />
  <!-- `key` can be any identifier for your source. -->
  <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
  <add key="contoso" value="https://contoso.com/packages/" />
</packageSources>

<!-- Define your source mapping element -->
<packageSourceMapping>
    <!-- Add a pattern for Microsoft.* or NuGet.* under the nuget.org source. -->
    <packagesource key="nuget.org">
        <package pattern="Microsoft.*" />
        <package pattern="NuGet.*" />
    </packageSource>
 
</packageSourceMapping>
```

Repeat this process for all sources.
When using package source mapping, every source should have a list of allowed package patterns.
A user may also pin a specific package id instead of broader package ID prefix.

> The most specific pattern that matches a package ID will have the highest precedence. Exact package IDs will always have the highest precedence over any prefix.

Reviewer: Maybe change: Exact package IDs will always have the highest precedence over any prefix.
 
to 

Package id patterns will always have the highest precedence over any prefix.

Author: ```suggestion
The most specific pattern that matches a package ID will have the highest precedence. Patterns with exact IDs will always have the highest precedence over a prefix pattern.
```

Reviewer: That's even better. 

Here's how I attempted to solve the problem in the feature docs: 

https://review.docs.microsoft.com/en-us/nuget/consume-packages/package-source-mapping?branch=dev-nkolev92-addPackageSourceMappingDocumentation

Figured a visual solution it's easier than any text :D

The global packages folder is an *append only* resource. This means NuGet only ever installs a new package. There is **no** refreshing, or overriding, or `re-installing` packages of any kind.
- When requesting a particular version (not floating), local, file sources are currently preferred - When installing a package, NuGet checks local sources independently before checking http sources.
- Package installation is operation based - If 3 projects are being restored during that operation, and all those projects have a dependency to `Newtonsoft.Json`, version `9.0.1`, in regular scenarios, only 1 project will *actually* download and install the package. The other projects will use the already installed version.
- `.nupkg.metadata` - Each package installation directory contains a `.nupkg.metadata` file which signify that a package installation is complete. This is expected to be the last file written in the package directory. This file is *not* use during build. NuGet.Client writes the package source information inside this file.

> #### Package Source Mapping rules

Reviewer: They were deleted, but we shoiuld keep the package installation rules. 

The doc contains both feature and technical aspects.

Author: AFAICT the rules I included kept all the relevant technical details from the previous rules. Here are the old ones:

> #### Package installation rules
> 
> - When the requested package is already installed in the global packages folder, no source look-up will happen. The namespaces are irrelevant.
> - Only sources with namespace metadata will be used.
> - When a source has no metadata, the default is *none*.
> - When a package id needs to be downloaded, it is first compared against sources with namespaces. If any match, only those sources are used. If none of the namespaces match, then the sources without namespaces are considered, if any.
> - A namespace with a specific package id is *always* preferred over a namespace with a prefix.
> - When multiple different namespaces match the package id, the most specific one will be considered.

Anything you see missing?

Reviewer: My focus was primarily on how the global packages folder is used as the source of truth. 
Should be better now.

> 1. Two types of package ID patterns are supported:

Reviewer: ```suggestion
1. Two types of package patterns are supported:
```

Reviewer: I see this matches the blog. 

I have some feedback on the particular parts, but I think we can do that in the docs instead of the spec. 
Up to you. Happy to provide the feedback.

<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->

## Future Possibilities

<!-- What future possibilities can you think of that this proposal would help with? -->
- Different modes or strategies can be considered in future iterations of the feature.
- The current proposal focuses on the nuget.config experience only. Adding a capability to source pin through RestoreSources is a future task.
- NuGet can allow users to filter their Package Source Mapping per source within CLI & IDE experiences.
- NuGet can allow a user to add a full or glob package ID namespace at install time with an additional click/parameter in Visual Studio or CLI.
- NuGet can combine this feature with `Package Lock Files` to allow a user to ensure the lock file is generated under the allowlist of package ID patterns.
- Create a better experience for this feature interacting with the global packages folder so that packages in the GPF are validated against the source mapping configuration.

> - Enable users to validate the outcome of their configuration with a CLI command, maybe apart of `dotnet list package`

Reviewer: ```suggestion
- Enable users to validate the outcome of their configuration with a CLI command, maybe a part of `dotnet list package`
```
